File: scope104/src/main/java/com/petro/scope104/data/db/dao/UserDao.java
package com.petro.scope104.data.db.dao;

import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import com.petro.scope104.data.db.entity.UserEntity;

import org.jetbrains.annotations.NotNull;

import java.util.List;

@Dao
public interface UserDao {
    @Query(
            "SELECT * " +
            "FROM user u " +
            "JOIN country c on c.countryCode = u.nat " +
            "WHERE (:isMale is NULL OR isMale = :isMale) AND (u.nat in (:nationality))" +
            "LIMIT :limit " +
            "OFFSET (:page * :limit) "
    )
    LiveData<List<UserEntity>> loadUsers(int page, int limit, @Nullable Boolean isMale, @NotNull List<String> nationality);

    @Query(
            "SELECT * " +
                    "FROM user u " +
                    "JOIN country c on c.countryCode = u.nat " +
                    "WHERE (:isMale is NULL OR isMale = :isMale)" +
                    "LIMIT :limit " +
                    "OFFSET (:page * :limit) "
    )
    LiveData<List<UserEntity>> loadUsers(int page, int limit, @Nullable Boolean isMale);


    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(List<UserEntity> user);
}

File: scope104/src/main/kotlin/com/petro/scope104/data/db/dao/UserDao.kt
package com.petro.scope104.data.db.dao

import androidx.room.*
import com.petro.scope104.data.db.entity.UserEntity
import io.reactivex.rxjava3.core.Observable

@Dao
interface UserDao {
    @Query(
        "SELECT * " +
                "FROM user u " +
                "JOIN country c on c.countryCode = u.nat " +
                "WHERE (:isMale is NULL OR isMale = :isMale) AND (u.nat in (:nationality))" +
                "LIMIT :limit " +
                "OFFSET (:page * :limit) "
    )
    fun loadUsers(
        page: Int,
        limit: Int,
        isMale: Boolean?,
        nationality: List<String?>
    ): Observable<List<UserEntity>>

    @Query(
        "SELECT * " +
                "FROM user u " +
                "JOIN country c on c.countryCode = u.nat " +
                "WHERE (:isMale is NULL OR isMale = :isMale)" +
                "LIMIT :limit " +
                "OFFSET (:page * :limit) "
    )
    fun loadUsers(page: Int, limit: Int, isMale: Boolean?): Observable<List<UserEntity>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insert(user: List<UserEntity?>?)
}